File: marketplace/marketplace-core/src/main/java/at/jku/cis/iVolunteer/marketplace/rule/DerivationRuleService.java
package at.jku.cis.iVolunteer.marketplace.rule;

import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import at.jku.cis.iVolunteer.marketplace.meta.core.property.definition.flatProperty.FlatPropertyDefinitionRepository;
import at.jku.cis.iVolunteer.marketplace.rule.engine.RuleService;
import at.jku.cis.iVolunteer.marketplace.rule.mapper.DerivationRuleMapper;
import at.jku.cis.iVolunteer.model.meta.core.property.definition.flatProperty.FlatPropertyDefinition;
import at.jku.cis.iVolunteer.model.rule.DerivationRule;
import at.jku.cis.iVolunteer.model.rule.engine.RuleExecution;
import at.jku.cis.iVolunteer.model.rule.entities.ClassActionDTO;
import at.jku.cis.iVolunteer.model.rule.entities.DerivationRuleDTO;

@Service
public class DerivationRuleService {

	@Autowired private DerivationRuleRepository derivationRuleRepository;
	@Autowired private DerivationRuleMapper derivationRuleMapper;
	@Autowired private FlatPropertyDefinitionRepository propertyDefinitionRepository;
	@Autowired private RuleService ruleService;
		
	public List<DerivationRuleDTO> getRules(String tenantId) {
		List<DerivationRule> all = derivationRuleRepository.getByTenantId(tenantId);
		return derivationRuleMapper.toTargets(all);
	}
	
	public DerivationRuleDTO getRule(String id) {
		return derivationRuleMapper.toTarget(derivationRuleRepository.findOne(id));
	}
	
	public DerivationRuleDTO getRuleByContainerAndName(String tenantId, String container, String ruleName) {
		return derivationRuleMapper.toTarget(derivationRuleRepository.getByTenantIdAndContainerAndName(tenantId, container, ruleName));
	}

	public DerivationRuleDTO createRule(DerivationRuleDTO derivationRuleDTO) {
		DerivationRule derivationRule = derivationRuleMapper.toSource(derivationRuleDTO);
		derivationRuleRepository.save(derivationRule);
		ruleService.addRule(derivationRule);
		//ruleService.executeRulesForAllVolunteers(derivationRule.getTenantId(), derivationRule.getContainer());
		return derivationRuleMapper.toTarget(derivationRule);
	}
	
	public List<RuleExecution> testRule(DerivationRuleDTO derivationRuleDTO) {
	    DerivationRule derivationRule = derivationRuleMapper.toSource(derivationRuleDTO);
		return ruleService.testRule(derivationRule);
	}

	public void updateRule(String id, DerivationRuleDTO derivationRuleDTO) {
		DerivationRule derivationRule = derivationRuleMapper.toSource(derivationRuleDTO);
		ruleService.addRule(derivationRule);
		derivationRuleRepository.save(derivationRule);
	}
	
	public List<FlatPropertyDefinition<Object>> getGeneralProperties(String tenantId){
		List<FlatPropertyDefinition<Object>> properties = new ArrayList<FlatPropertyDefinition<Object>>();
		
		FlatPropertyDefinition<Object> pd = (FlatPropertyDefinition<Object>) propertyDefinitionRepository.getByNameAndTenantId("Alter", tenantId).get(0);
		properties.add(pd);
		
		return properties;
	}
	
}

File: marketplace-frontend/src/app/main/content/_service/derivation-rule.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { isNullOrUndefined } from 'util';
import { Marketplace } from '../_model/marketplace';
import { DerivationRule } from '../_model/derivation-rule';

@Injectable({
  providedIn: 'root'
})
export class DerivationRuleService {

  constructor(private http: HttpClient) {
  }

  findByIdAndTenantId(marketplace: Marketplace, id: string, tenantId: string){
    return this.http.get(`${marketplace.url}/rule/${id}/tenant/${tenantId}`);
  }

  findById(marketplace: Marketplace, id: string) {
    return this.http.get(`${marketplace.url}/rule/${id}`);
  }

  findByContainerAndName(marketplace: Marketplace, tenantId: string, container: string, ruleName: string){
    return this.http.get(`${marketplace.url}/rule/tenant/${tenantId}/container/${container}/rule/${ruleName}`);
  }

  findAll(marketplace: Marketplace, tenantId: string) {
    return this.http.get(`${marketplace.url}/rule/tenant/${tenantId}`);
  }

  test(marketplace: Marketplace, derivationRule: DerivationRule){
    return this.http.post(`${marketplace.url}/rule/test`, derivationRule);
  }

  getTestResults(marketplace: Marketplace, derivationRule: DerivationRule){
    return this.http.get(`${marketplace.url}/rule/test/${derivationRule.container}`);
  }

  save(marketplace: Marketplace, derivationRule: DerivationRule) {
    if (isNullOrUndefined(derivationRule.id)) {
      return this.http.post(`${marketplace.url}/rule`, derivationRule);
    }
    return this.http.put(`${marketplace.url}/rule/${derivationRule.id}`, derivationRule);
  }

  getGeneralProperties(marketplace: Marketplace, tenantId: string){
    return this.http.get(`${marketplace.url}/rule/tenant/${tenantId}/general/properties`);
  }
}
